internal/controllers/machinedeployment/machinedeployment_sync.go

```diff
@@ -86,173 +89,277 @@ func (r *Reconciler) getAllMachineSetsAndSyncRevision(ctx context.Context, d *cl
 		err = r.updateMachineDeployment(ctx, d, func(innerDeployment *clusterv1.MachineDeployment) {
-			mdutil.SetDeploymentRevision(d, msCopy.Annotations[clusterv1.RevisionAnnotation])
+			mdutil.SetDeploymentRevision(d, updatedMS.Annotations[clusterv1.RevisionAnnotation])
 		})
-		return msCopy, err
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to update revision annotation on MachineDeployment")
+		}
+		return updatedMS, nil
 	}
 
-	if !createIfNotExisted {
+	if !createIfNotExists {
 		return nil, nil
 	}
 
-	// new MachineSet does not exist, create one.
-	newMSTemplate := *d.Spec.Template.DeepCopy()
-	hash, err := mdutil.ComputeSpewHash(&newMSTemplate)
+	// Create a new MachineSet and wait until the new MachineSet exists in the cache.
+	newMS, err := r.createMachineSetAndWait(ctx, d, oldMSs)
 	if err != nil {
 		return nil, err
 	}
-	machineTemplateSpecHash := fmt.Sprintf("%d", hash)
-	newMSTemplate.Labels = mdutil.CloneAndAddLabel(d.Spec.Template.Labels,
-		clusterv1.MachineDeploymentUniqueLabel, machineTemplateSpecHash)
 
-	// Add machineTemplateHash label to selector.
-	newMSSelector := mdutil.CloneSelectorAndAddLabel(&d.Spec.Selector,
-		clusterv1.MachineDeploymentUniqueLabel, machineTemplateSpecHash)
-
-	minReadySeconds := int32(0)
-	if d.Spec.MinReadySeconds != nil {
-		minReadySeconds = *d.Spec.MinReadySeconds
+	// Ensure MachineDeployment has the latest MachineSet revision in its revision annotation.
+	err = r.updateMachineDeployment(ctx, d, func(innerDeployment *clusterv1.MachineDeployment) {
+		mdutil.SetDeploymentRevision(d, newMS.Annotations[clusterv1.RevisionAnnotation])
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to update revision annotation on MachineDeployment")
 	}
 
-	// Create new MachineSet
-	newMS := clusterv1.MachineSet{
-		ObjectMeta: metav1.ObjectMeta{
-			// Make the name deterministic, to ensure idempotence
-			Name:      d.Name + "-" + apirand.SafeEncodeString(machineTemplateSpecHash),
-			Namespace: d.Namespace,
-			Labels:    make(map[string]string),
-			// Note: by setting the ownerRef on creation we signal to the MachineSet controller that this is not a stand-alone MachineSet.
-			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(d, machineDeploymentKind)},
-		},
-		Spec: clusterv1.MachineSetSpec{
-			ClusterName:     d.Spec.ClusterName,
-			Replicas:        new(int32),
-			MinReadySeconds: minReadySeconds,
-			Selector:        *newMSSelector,
-			Template:        newMSTemplate,
-		},
+	return newMS, nil
+}
+
+// updateMachineSet updates an existing MachineSet to propagate in-place mutable fields from the MachineDeployment.
+func (r *Reconciler) updateMachineSet(ctx context.Context, deployment *clusterv1.MachineDeployment, ms *clusterv1.MachineSet, oldMSs []*clusterv1.MachineSet) (*clusterv1.MachineSet, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Compute the desired MachineSet.
+	updatedMS, err := r.computeDesiredMachineSet(deployment, ms, oldMSs, log)
```

Comment:
Am I right to say that in-place fields won't propagate to old MachineSets/Machines? This might be surprising UX, e.g.
- User change aws template ref AND drain timeout.
- New MachineSet is created with new drain timeout.
- Old MachineSet perform progressive scale down honouring old drain timeout.

We'll need to documented and clarify all these expectations/scenarios

Reply:
> Am I right to say that in-place fields won't propagate to old MachineSets/Machines?

Yes. In-place propagation only propagates into our current / new / target MachineSet. It doesn't touch old MachineSets and their corresponding Machines.

Good point about documenting expectations.

Reply:
Comment was moved over to https://github.com/kubernetes-sigs/cluster-api/pull/8110#discussion_r1106327497